### PR TITLE
Stream output file to stdout

### PR DESCRIPTION
**Pull request by Meakk**
Implement #1101

### CODE REVIEW

**Comment by Meakk** [comment]
@mwestphal @snoyer please take a look :)

File: application/F3DStarter.cxx [hunk 1]
```diff
@@ -164,8 +165,16 @@ int F3DStarter::Start(int argc, char** argv)
...
+  const bool renderToStdout = this->Internals->AppOptions.Output == "-";
+  if (renderToStdout)
+  {
+    f3d::log::setVerboseLevel(f3d::log::VerboseLevel::ERROR);
+  }
+  else
+  {
+    // Set verbosity level early from command line
+    F3DInternals::SetVerboseLevel(this->Internals->AppOptions.VerboseLevel);
+  }
```
Comment (snoyer) [inline application/F3DStarter.cxx:177]:
```suggestion
  std::streambuf* originalStdout = std::cout.rdbuf();
  if (renderToStdout)
  {
    std::cout.rdbuf(std::cerr.rdbuf());
  }

  // Set verbosity level early from command line
  F3DInternals::SetVerboseLevel(this->Internals->AppOptions.VerboseLevel);
```

File: application/F3DStarter.cxx [hunk 2]
```diff
@@ -426,7 +438,15 @@ int F3DStarter::Start(int argc, char** argv)
...
+      if (renderToStdout)
+      {
+        const auto buffer = img.saveBuffer();
+        std::copy(buffer.begin(), buffer.end(), std::ostreambuf_iterator(std::cout));
+      }
+      else
+      {
+        img.save(this->Internals->AppOptions.Output);
+      }
       f3d::log::debug("Output image saved to ", this->Internals->AppOptions.Output);
```
Comment (snoyer) [inline application/F3DStarter.cxx:450]:
```suggestion
      if (renderToStdout)
      {
        const auto buffer = img.saveBuffer();
        std::copy(buffer.begin(), buffer.end(), std::ostreambuf_iterator(originalStdout));
        f3d::log::debug("Output image data printed to stdout");
      }
      else
      {
        img.save(this->Internals->AppOptions.Output);
        f3d::log::debug("Output image saved to ", this->Internals->AppOptions.Output);
      }
```

**Review by snoyer — COMMENTED** [review]
The key point here is that we want to ensure _nothing_ has been written to `stdout` before we use it to output render data, otherwise we will quietly output a corrupted PNG.

Setting the log level to `ERROR` early seemed to do a good enough job when I suggested it but I haven't fully tested it.
Possible issues include:
- if `ERROR` level messages go to `stdout` _and_ can we have non-critical error messages (ie. not exceptions and we still proceed to render)
- if there's any code that output to `stdout` directly (most likely external code not going through `f3d::log`)
- if there are cases where you'd want to render to stdout _and_ still see log messages somehow

Maybe it'd be better to not mess with the verbosity levels and instead _redirect_ `stdout` to `stderr` from the start of the app if we know we'll be rendering to stdout?
That way any logging would still get to the terminal through `stderr` and only the PNG data would go to `stdout` (`f3d foo.bar --verbose --output=- > /tmp/f3d.png` would output both the full log and a valid PNG).
Would that be just a dirty hack? Are there any critical implications I'm not seeing?

File: application/F3DStarter.cxx (hunk 1 above)
Comment (mwestphal) [reply application/F3DStarter.cxx:177]: interesting suggestion, how do other cli software do it ?

Comment (Meakk) [reply application/F3DStarter.cxx:177]: Honestly I think we shouldn't even do anything about the verbose level automatically. Let's let the user specify the right verbosity level manually?
About external libraries potentially writing to std::cout, it's not our problem and we can still provide patches upstream if needed?

Comment (Meakk) [reply application/F3DStarter.cxx:177]: And the default verbosity behavior is only writing in std::cerr anyway?

Comment (snoyer) [reply application/F3DStarter.cxx:177]: > Honestly I think we shouldn't even do anything about the verbose level automatically.

That part definitely,  let's not change it magically.

> Let's let the user specify the right verbosity level manually?
> About external libraries potentially writing to std::cout, it's not our problem and we can still provide patches upstream if needed?

These 2 I guess can be argued both ways, we could do nothing and let the users find out if their output get corrupted, then either tell them they did it wrong in the first case, or go patch and rebuild in the second case.
Redirecting would be the "safe" way but I can't say if it's a good/nice/clean thing to do, or if anybody else does it that way as @mwestphal asked...

> And the default verbosity behavior is only writing in std::cerr anyway?

I haven't checked so I'll trust you on that, in which case it's fine as long as we trust libraries (looking at you OCCT...) to not randomly print directly.

Comment (Meakk) [reply application/F3DStarter.cxx:177]: Redirect to stderr should be easy. We just have to override that: vtkOutputWindow::GetDisplayStream
So let's decide. I'm not against the idea.

Comment (mwestphal) [reply application/F3DStarter.cxx]: that sounds like a good idea to me.

File: library/public/log.h [hunk 3]
```diff
@@ -101,8 +101,10 @@ class F3D_EXPORT log
 
   /**
    * Set the verbose level.
+   * By default, only warnings and errors are written to stderr, debug and info are written to
+   * stdout. If forceStdErr is true, all messages including debug and info are written to stderr.
    */
-  static void setVerboseLevel(VerboseLevel level);
+  static void setVerboseLevel(VerboseLevel level, bool forceStdErr = false);
```
Comment (mwestphal) [inline library/public/log.h:107]: Does adding another staticc method makes sense ?

Comment (Meakk) [reply library/public/log.h:107]: That was my initial plan but it is calling the same function so exposing the display stream mode makes more sense imho

Comment (mwestphal) [reply library/public/log.h:107]: right I get your point.

**Review by mwestphal — APPROVED** [review]

**Pull request merged**